File: xdm-core/src/main/java/xdm/core/media/parser/dash/PeriodParser.java
//package xdm.core.media.parser.dash;
//
//import java.net.URI;
//import java.util.ArrayList;
//import java.util.List;
//import org.w3c.dom.Node;
//import org.w3c.dom.NodeList;
//
//public class PeriodParser {
//  private PeriodParser() {}
//
//  public static List<MpdEntry> parsePeriod(
//      Node period, URI baseUrl, long mediaPresentationDuration) {
//    long periodDuration = mediaPresentationDuration;
//    String durationAttr = XmlUtil.getAttr(period, "duration");
//    if (durationAttr != null) {
//      periodDuration = DashUtil.parseXsDuration(durationAttr);
//    }
//    baseUrl = XmlUtil.resolveBaseUrl(period, baseUrl);
//    boolean hasParentSegmentBase = XmlUtil.containsTag(period, "SegmentBase");
//    NodeList adaptationSets = ((org.w3c.dom.Element) period).getElementsByTagName("AdaptationSet");
//    List<Representation> audioList = new ArrayList<>();
//    List<Representation> videoList = new ArrayList<>();
//    for (int i = 0; i < adaptationSets.getLength(); i++) {
//      List<Representation> representations =
//          RepresentationParser.parseAdaptationSet(
//              adaptationSets.item(i), baseUrl, periodDuration, hasParentSegmentBase);
//      if (representations.isEmpty()) continue;
//      if (representations.get(0).getMimeType().startsWith("audio")) {
//        audioList.addAll(representations);
//      } else {
//        videoList.addAll(representations);
//      }
//    }
//    List<MpdEntry> mediaList = new ArrayList<>();
//    if (!videoList.isEmpty() && !audioList.isEmpty()) {
//      for (Representation video : videoList) {
//        for (Representation audio : audioList) {
//          mediaList.add(new MpdEntry(video, audio));
//        }
//      }
//    } else if (!videoList.isEmpty()) {
//      for (Representation video : videoList) {
//        mediaList.add(new MpdEntry(video, null));
//      }
//    } else if (!audioList.isEmpty()) {
//      for (Representation audio : audioList) {
//        mediaList.add(new MpdEntry(null, audio));
//      }
//    }
//    return mediaList;
//  }
//}

File: xdm-core/src/main/java/xdm/core/downloaders/web/streaming/manifest/dash/PeriodParser.kt
package xdm.core.downloaders.web.streaming.manifest.dash

import org.w3c.dom.Element
import org.w3c.dom.Node
import java.net.URI


fun parsePeriod(
    period: Node, baseUrl: URI?, mediaPresentationDuration: Long
): List<MpdEntry> {
    var baseUrl = baseUrl
    var periodDuration = mediaPresentationDuration
    val durationAttr = getAttr(period, "duration")
    if (durationAttr != null) {
        periodDuration = parseXsDuration(durationAttr)
    }
    baseUrl = resolveBaseUrl(period, baseUrl!!)
    val hasParentSegmentBase = containsTag(period, "SegmentBase")
    val adaptationSets = (period as Element).getElementsByTagName("AdaptationSet")
    val audioList: MutableList<Representation> = ArrayList()
    val videoList: MutableList<Representation> = ArrayList()
    for (i in 0..<adaptationSets.length) {
        val representations = parseAdaptationSet(
                adaptationSets.item(i), baseUrl, periodDuration, hasParentSegmentBase
            )
        if (representations.isEmpty()) continue
        if (representations[0].mimeType.startsWith("audio")) {
            audioList.addAll(representations)
        } else {
            videoList.addAll(representations)
        }
    }
    val mediaList: MutableList<MpdEntry> = ArrayList()
    if (videoList.isNotEmpty() && audioList.isNotEmpty()) {
        for (video in videoList) {
            for (audio in audioList) {
                mediaList.add(MpdEntry(video, audio))
            }
        }
    } else if (videoList.isNotEmpty()) {
        for (video in videoList) {
            mediaList.add(MpdEntry(video, null))
        }
    } else if (audioList.isNotEmpty()) {
        for (audio in audioList) {
            mediaList.add(MpdEntry(null, audio))
        }
    }
    return mediaList
}
